Proposed bindings

dlecocq opened the pull request:
As discussed with @evanbattaglia and @b4hand , here are the bindings I've been hacking on.

I still have to update the README, TODOs

b4hand commented on README.md:
```diff
@@ -6,106 +8,295 @@
+- Have an exception hierarchy (Qless exceptions on down)
+- Worker don't fail job on lost lock exception
+- Set the redis instance for testing purposes
+- Add a --verbose flag for more logging. Also -vv
```
Isn't this done above?

dlecocq replied:
I did document the flag, but I haven't yet implemented. Carts and horses.

dlecocq replied:
This flag is now implemented.

b4hand commented on lib/job.js:
```diff
@@ -1,106 +1,249 @@
+ * Split the provided name by `/`, and import the longest working prefix. The
+ * remaining components will be treated as attributes off of the import.
+ *
+ * For example, consider 'some/package/file.js/property/class'. We would
```
This seems a little hokey to me. Is there a reason to not just look for the `*.js` extension? I don't feel strongly about it, and it appears to work, so maybe it's not a big deal, but it does seem a little complex.

dlecocq replied:
That's a fair question, and honestly, this setup leaves a bad taste in my mouth.

However, in general, installed modules do not have the `*.js`. I thought about having a special delimiter for what is the `properties` portion and what's the `path` portion, (`some/package/file.js:property.class`), but that seemed even hokier.

Now's the time to get this right, and I'm very open to suggestions. I'm not in love with this scheme either, and it's mostly just a placeholder.

b4hand commented on README.md:
```diff
@@ -6,106 +8,295 @@
+- Add a --verbose flag for more logging. Also -vv
+- Use multiple cores (until then, it's recommended to use parameterized upstart)
+- Support the `--resume` option
+- Recurring job objects
```
It looks like there is at least partial functionality for recurring jobs. Could you expound on what is yet to be done?

dlecocq replied:
Sure thing. While we can create recurring jobs, there is no corresponding `RecurringJobs` class to manipulate them. So while I can do this:

```javascript
client.job('my-normal-jid').then((job) => { job.move('another-queue'); })
```

I cannot yet do this:

```javascript
client.job('my-recurring-jid').then((job) => { job.move('another-queue') })
```

b4hand commented on package.json:
```diff
@@ -1,51 +1,59 @@
-  "homepage": "https://github.com/seomoz/qless-js#readme",
-  "main": "qless.js",
+  "author": "Moz",
+  "license": "UNLICENSED",
```
MIT?

dlecocq replied:
Good call. Copypasta from another package.

evanbattaglia commented on lib/workers/multi.js:
```diff
@@ -0,0 +1,57 @@
+        });
+
+      // Try to run another job, waiting on the semaphore.
+      return self.run();
```
clever

dlecocq commented:
Ok, I've managed to get it working for Node 4, 5, 6.

I also got the `-v` and `-vv` modes working, as well as the optional `--allow-paths` option in the binary, allowing Job `klass`es to be paths (rather than modules).

dlecocq commented on bin/qless-js-worker.js:
```diff
@@ -0,0 +1,36 @@
+const assert = require('assert');
+const qless = require('qless-js');
+
+var commander = require('commander');
```
Whoops, my bad. I wonder why my linter didn't catch this.

dlecocq commented on bin/qless-js-worker.js:
```diff
@@ -0,0 +1,36 @@
+  .option('-d, --workdir <dir>', 'The base work directory path')
+  .option('-c, --concurrency <n>', 'The number of concurrent jobs to run [default 10]')
+  .option('-i, --interval <n>', 'Polling interval in seconds [default 60]', parseInt)
+  .option('-q, --queue <name>', 'Add a queue to work on', collect, [])
```
D'oh. My linting invocation needs some adjustment.

dlecocq replied:
@evanbattaglia -- My linter doesn't mind the missing semicolon on this line. Do y'all have a `.eslintrc` config you'd like me to use? I'm happy to adhere to whatever style, so long as it it's lint-configured.

evanbattaglia replied:
oops, I replied in email and said "You can use the one I previously had... It's pretty similar to what we used in other projects on our team." -- but the context that this was referring to .eslintrc was lost.

evanbattaglia commented:
You can use the one I previously had... It's pretty similar to what we used
in other projects on our team.

2017-03-29 13:37 GMT-07:00 Dan Lecocq <notifications@github.com>:

> *@dlecocq* commented on this pull request.
> ------------------------------
>
> In bin/qless-js-worker.js
> <https://github.com/seomoz/qless-js/pull/11#discussion_r108780723>:
>
> > +const qless = require('qless-js');
> +
> +var commander = require('commander');
> +
> +function collect(name, queues) {
> +  queues.push(name);
> +  return queues;
> +}
> +
> +commander
> +  .option('-h, --host <url>', 'The redis:// url to connect to')
> +  .option('-n, --name <name>', 'The name to identify your worker as')
> +  .option('-d, --workdir <dir>', 'The base work directory path')
> +  .option('-c, --concurrency <n>', 'The number of concurrent jobs to run [default 10]')
> +  .option('-i, --interval <n>', 'Polling interval in seconds [default 60]', parseInt)
> +  .option('-q, --queue <name>', 'Add a queue to work on', collect, [])
>
> @evanbattaglia <https://github.com/evanbattaglia> -- My linter doesn't
> mind the missing semicolon on this line. Do y'all have a .eslintrc config
> you'd like me to use? I'm happy to adhere to whatever style, so long as it
> it's lint-configured.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/seomoz/qless-js/pull/11#discussion_r108780723>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAo2NC3tqugFjJXTImvlXzsUxdVx00AVks5rqsEUgaJpZM4Mq2Kq>
> .
>

evanbattaglia commented:
I'm not sure node has such a clear distinction between paths and modules. I
guess you could say a module is directory with a "package.json" -- in that
case node knows to look at the "main" js file listed in the package.json.
But you can still do things like `require("mymodule/lib/myfile")`, and if
`node_modules/mymodule/lib/myfile.js` exists it will pick it up. Also note
that the error message when you say `require("./doesnotexist.js")` is still
`Error: cannot find module './doesnotexist.js'`

2017-03-29 13:30 GMT-07:00 Dan Lecocq <notifications@github.com>:

> Ok, I've managed to get it working for Node 4, 5, 6.
>
> I also got the -v and -vv modes working, as well as the optional
> --allow-paths option in the binary, allowing Job klasses to be paths
> (rather than modules).
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/seomoz/qless-js/pull/11#issuecomment-290216291>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAo2NHqLc2SNAC6MTnJs0yesNkTTJpDWks5rqr9_gaJpZM4Mq2Kq>
> .
>

dlecocq commented on lib/client.js:
```diff
@@ -1,55 +1,131 @@
-    this.jobs = new jobs.Jobs(this);
-    this.workerName = [os.hostname(), process.env.HOST, process.pid.toString()].join('-');
-  }
+function Client(config) {
```
Old habits. I'll convert all these.

dlecocq commented on README.md:
````diff
@@ -6,106 +8,295 @@
 ```
 
-#### Example job with co/generators
+__Note__: A job class cannot be a path, by default. There is a [to-do](#to-dos)
````
WRT the leaving the `node_modules` directory, we defend against that by first `Path.resovle`-ing the provided `klass` and then checking if that begins with a `.` or `/`, and there's a test case about exactly the example you provided.

As for how this will work in an application, take this with a grain of salt as I don't have a lot of experience with deploying node applications. I imagine that in some form or another one will have a sandboxed environment (I imagine / hope there's something akin to `virtualenv` or `bundle` in the Node community -- I don't know yet if `yarn` meets those criteria, but that's what I'd been using).

And in that environment, I imagine application code and `qless-js` are installed and readily importable. That you'd make jobs with a `klass` like `my-application/job` or the like, and you just run a `qless-js-worker` in upstart in the context of that environment?

My intention is to do as is done in the Node community -- the above doesn't necessarily express a die-hard opinion, but more what I have in mind with the understanding that there may be some gaps in my understanding :-)

evanbattaglia replied:
OK, I missed the `Path.resolve`.

The projects I've seen usually use relative requires. There are things like webpack which will resolve `import`s starting from a certain path, but I don't think that applies here. Setting `NODE_PATH` to the `app`/`src`/etc. directory should work, but I don't think everyone automatically sets that, so you should mention it in the docs.

dlecocq replied:
I mean, this still allows relative imports from _within_ a package or module. It just means that the initial import should be of an installed module. For instance:

```javascript
// my-module
module.exports = {
  Job: require('./lib/job.js'),
}
```

That would work as expected when specifying a job `my-module/Job`.

dlecocq commented on test/client.test.js:
```diff
@@ -0,0 +1,88 @@
+const helper = require('./helper.js');
+
+describe('Qless client', () => {
+  let client = new Client();
```
Yeah, I'm all for that. Addressing all the linting issues now.

dlecocq commented on lib/job.js:
```diff
@@ -1,106 +1,249 @@
+const Path = require('path');
+const logger = require('./logger.js');
+
+function Job(client, config) {
```
Yup. I'll be `class`-ing all the things. For whatever reason, I thought `class` was only available in >6, but I guess I was wrong.

dlecocq commented on lib/job.js:
```diff
@@ -1,106 +1,249 @@
+ *  - import 'some/package/file.js', get 'property.class' attribute (success)
  */
+Job.import = function(name) {
+  let separator = '/';
```
Fair

dlecocq replied:
There ended up being such a value defined in `Path.sep`.

dlecocq commented on lib/job.js:
```diff
@@ -1,106 +1,249 @@
-    return klassFinder.findClass(this.klassName);
-  }
+Job.prototype.setPriority = function(priority) {
+  let self = this;
```
So... about the `self` / `this` stuff -- in my testing, that was definitely breaking things. I know that there's a `bluebird` `bind`, which may be in order. In either case, I'll check it out.

evanbattaglia replied:
I don't know why it wasn't working for you, but `=>` definitely binds.

```
class Foo {
  constructor(x) {
    this.x = x;
  }

  foo() {
    return new Promise(res => res(4)).then(val => val + this.x);
  }

}

const f = new Foo(2);
f.foo().then(console.log); // 6
```

dlecocq replied:
I don't know what I was doing wrong before, but I was able to get rid of all the `self`s. Bogons.

dlecocq commented:
Re: the distinction between paths and modules, for me it just means whether it's installed somewhere in my node module path. In normal application mode, I only want to be able to run jobs that are installed somewhere, rather than just some random path on the system.

evanbattaglia commented on .travis.yml:
```diff
@@ -7,9 +7,8 @@ node_js:
 script:
-  - npm run coverage
-  - npm run lint
+  - yarn run test
```
Never mind, I had never heard of it but I read up on it and it seems OK

dlecocq commented:
By way of an update, I've been working today to resolve the linting errors after I got my linter configured properly. D'oh.

I'm in the middle of rebasing and addressing all the feedback, and expect to have an updated PR tomorrow.

dlecocq commented on lib/client.js:
```diff
@@ -1,55 +1,131 @@
-  }
+// Return a new disposer
+Client.disposer = function(config) {
+  return new bluebird.Promise((resolve, reject) => {
```
Ended up with `return Promise.resolve(new Client(config)).disposer(client => client.quit());`

And yes, the linting _did_ complain once I got the invocation right.

dlecocq commented on lib/client.js:
```diff
@@ -1,55 +1,131 @@
+      } else {
+        // Try to get a recurring job
+        return self.call('recur.get', jid)
+          .then((job) => {
```
Yeah, I agree. The linting complained about the reuse of the name `job`, so I ended up going with `recurringJob`, which made the single line a bit unwieldy.

dlecocq commented on README.md:
```diff
@@ -6,106 +8,295 @@
+[Job Class Importing](#job-class-importing) section below.
+
+The function that gets run is `<job class>.<queue name>`. So our example will
+trun a `MyClass.myQueue` function, passing the popped job. This function:
```
done

dlecocq commented on README.md:
````diff
@@ -6,106 +8,295 @@
 
 ```bash
-DEBUG='qless:*' node worker.js
+qless-js-worker \
````
Executable works for me

dlecocq replied:
WRT installation, my expectation is that it would be installed in the environment in which it's deployed. Be that globally or `yarn`

dlecocq commented on lib/job.js:
```diff
@@ -1,106 +1,249 @@
+ */
+Job.prototype.getKlass = function() {
+  let name = Path.normalize(this.klassName);
+  if (name.startsWith('.') || name.startsWith('/')) {
```
Just to close the loop here, too, because of the `Path.normalize`, we can check for an indication of being an absolute path (starting with `/`) or a relative path (starting with `.`). Everything else would be interpreted as relative to `node_modules`.

dlecocq commented on lib/workers/popper.js:
```diff
@@ -0,0 +1,70 @@
+          }
+        }
+
+        if (job) {
```
👍

dlecocq commented on lib/workers/single.js:
```diff
@@ -0,0 +1,46 @@
+          return;
+        }
+      })
+  })
```
The linting wants all branches of functions to return, so it'd have to be a ternary operator, which starts to get a little long. Even so, I did incorporate your suggestion about the invocation to `this.run`

dlecocq commented:
I believe now all the comments have been addressed, feedback incorporated.

@b4hand @evanbattaglia -- would you guys be willing to take another look? No rush.

b4hand commented on README.md:
```diff
@@ -6,106 +8,297 @@
+
+- Sandbox work directory
+- Have an exception hierarchy (Qless exceptions on down)
+- Worker don't fail job on lost lock exception
```
s/Worker/Workers/

dlecocq replied:
noted

b4hand commented on bin/qless-js-worker.js:
```diff
@@ -0,0 +1,51 @@
+}
+
+commander
+  .option('-h, --host <url>', 'The redis:// url to connect to')
```
I wonder if this would be clearer if it were specified as `--redis` instead of `--host`. I know that probably follows the pattern of other qless workers, but I wonder if this was an original oversight that should be remedied.

dlecocq replied:
I think you're right.

b4hand commented on lib/job.js:
```diff
@@ -1,106 +1,263 @@
-    if (_.isEmpty(this.dependencies)) this.dependencies = [];
+        logger.debug('Found "%s"', result.name);
+        return result;
+      } catch (err) {
```
I think it would be more clear if this `try/catch` were just around the `require` since I believe that's all it's trying to catch.

dlecocq replied:
That's a good point.

b4hand commented on lib/queue.js:
```diff
@@ -1,58 +1,216 @@
+   *  - depends (array of string jids on which this job depends)
+   */
+  put(options) {
+    // TODO(dan): should this yell if you provide an unsupported option?
```
Regarding this comment, I would ask "how often has the job format changed, and do you want to allow independent evolution of the job format across language bindings?"

b4hand commented on lib/queue.js:
```diff
@@ -1,58 +1,216 @@
+
+  /**
+   * Return the current statistics for a given queue on a given date.
+   * The results are returned are a JSON blob::
```
s/The results are returned are a JSON blob::/The results returned are a JSON blob:/

dlecocq commented:
Thanks, @evanbattaglia for your thorough review -- I really appreciate it and I learned several things from it!

dlecocq commented:
I'm going to go ahead and merge. It's been through a couple iterations, and any future feedback can be applied retroactively.
